File: Leetcode/307_2.cpp
/**
 * 307. 区域和检索 - 数组可修改 线段树(数组实现)
 * link：https://leetcode-cn.com/problems/range-sum-query-mutable/
 */

class NumArray
{
    vector<int> store;
    int n;

public:
    NumArray(vector<int> &nums)
    {
        n = nums.size();
        store.resize(2 * n);

        for (int i = 0; i < n; i++)
            store[i + n] = nums[i];
        for (int i = n - 1; i > 0; i--)
            store[i] = store[2 * i] + store[2 * i + 1];
    }

    void update(int i, int val)
    {
        int pos = i + n;
        int diff = val - store[pos];

        store[pos] = val;

        while (pos > 0)
        {
            store[pos / 2] += diff;
            pos /= 2;
        }
    }

    int sumRange(int i, int j)
    {
        i += n;
        j += n;
        int sum = 0;
        while (i <= j)
        {
            if (i % 2 == 1)
            {
                sum += store[i];
                i++;
            }
            if (j % 2 == 0)
            {
                sum += store[j];
                j--;
            }
            i /= 2;
            j /= 2;
        }
        return sum;
    }
};

/**
 * Your NumArray object will be instantiated and called as such:
 * NumArray* obj = new NumArray(nums);
 * obj->update(i,val);
 * int param_2 = obj->sumRange(i,j);
 */

File: Leetcode/307_2.py
# 307. 区域和检索 - 数组可修改 线段树(数组实现)
# link：https://leetcode-cn.com/problems/range-sum-query-mutable/

class NumArray:
    def __init__(self, nums: List[int]):
        self.n = len(nums)
        self.store = [0] * self.n + nums

        for i in range(self.n-1, 0, -1):
            self.store[i] = self.store[2 * i] + self.store[2 * i + 1]

    def update(self, i: int, val: int) -> None:
        i += self.n

        diff: int = val - self.store[i]
        while i > 0:
            self.store[i] += diff
            i = i // 2

    def sumRange(self, i: int, j: int) -> int:
        i += self.n
        j += self.n
        sum: int = 0

        while i <= j:
            if i % 2 == 1:
                sum += self.store[i]
                i += 1
            if j % 2 == 0:
                sum += self.store[j]
                j -= 1
            i = i // 2
            j = j // 2

        return sum
